Airbrakes/airbrakes.py

```diff
@@ -38,19 +38,29 @@ def __init__(self, logger: Logger, servo: Servo, imu: IMU):
 
     def update(self):
         """
-        Called every loop iteration. Depending on the current state, it will do different things. It
-        is what controls the airbrakes and chooses when to move to the next state.
+        Called every loop iteration from the main process. Depending on the current state, it will
+        do different things. It is what controls the airbrakes and chooses when to move to the next
+        state.
         """
         # Gets the current extension and IMU data, the states will use these values
         self.current_extension = self.servo.current_extension
-        self.current_imu_data = self.imu.get_imu_data()
+        data_packets = []
+
+        # Let's get 50 data packets to ensure we have enough data to work with.
+        # 50 is an arbitrary number for now - if the time resolution between each data packet is
+        # 2ms, then we have 2*50 = 100ms of data to work with at once.
+        while len(data_packets) < 50:
```

Comment:
I don't know if I love this -- I don't like that how fast the system runs is linked with how fast/how much data it is given. This should work and *should* be fine, but I don't think its good practice. It also feels wrong having the context hold onto the data packets.

I think I would rather the imu class hold onto its most recent data packets. Maybe in this update we have a method first check if there is any data, then if there is, pop all of it off and log it, and then have the states use the most recent data point?

This is a loose idea and it very well might have a lot of flaws, so lmk what you think

Reply:
> I don't know if I love this -- I don't like that how fast the system runs is linked with how fast/how much data it is given. This should work and _should_ be fine, but I don't think its good practice. 

Yeah, I did this because we need to process data over time, in real time. Technically, if the IMU is sending erroneous data or very delayed data, we are cooked and can't do anything anyway. 

I just got an idea on how to improve this - We could calculate the sampling rate (via e.g. moving average) of the IMU in real time, so if it drops below a certain threshold we can log that / error out / actually replace this "50" number and calculate a good length. That's of course a tiny performance overhead, but it should be worth the data and insight?

> It also feels wrong having the context hold onto the data packets. I think I would rather the imu class hold onto its most recent data packets

Agreed, that it should be abstracted away by the `IMU` class.

> and then have the states use the most recent data point?

but that's the thing, don't we want to calculate the state update, and that needs a couple of data packets? How can we determine the next state by looking at a single data point?